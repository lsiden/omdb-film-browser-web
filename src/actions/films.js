/*
 * I discovered that OMDB search can return some duplicate films.
 * This will cause React to throw a warning when it discovers a duplicate key.
 * So we have to guard against duplicate items.
 */

const reduceUniqueFilms = (uniqueIds, res, film) => {
  const id = film.imdbID

  if (!uniqueIds[id]) {
    res.push(film)
    uniqueIds[id] = id
  }
  return res
}

export function updateNewFilms(state, data) {
  const uniqueIds = {}
  const newFilms = data.films || []
  return {
    ...state,
    ...data,
    uniqueIds,
    films: newFilms.reduce(reduceUniqueFilms.bind(null, uniqueIds), []),
  }
}

export function appendFilms(state, data) {
  const newFilms = data.films || []
  const prevFilms = state.films || []
  const uniqueIds =
    state.uniqueIds ||
    prevFilms.reduce((res, film) => {
      const id = film.imdbID

      if (!res[id]) {
        res[id] = id
      }
      return res
    }, {})
  return {
    ...state,
    ...data,
    uniqueIds,
    films: newFilms.reduce(reduceUniqueFilms.bind(null, uniqueIds), [
      ...prevFilms,
    ]),
  }
}
