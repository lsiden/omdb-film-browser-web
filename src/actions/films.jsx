/*
 * I discovered that OMDB search can return some duplicate films.
 * This will cause React to throw a warning when it discovers a duplicate key.
 * So we have to guard against duplicate items.
 */

// FIXME - try to get rid of uniqueIds

const reduceUniqueFilms = (uniqueIds, res, film) => {
  const id = film.imdbID

  if (!uniqueIds[id]) {
    res.push(film)
    uniqueIds[id] = id /* eslint-disable-line no-param-reassign */
  }
  return res
}

export function updateNewFilms(state, data) {
  const uniqueIds = {}
  const newFilms = data.films || []
  const films = newFilms.reduce(reduceUniqueFilms.bind(null, uniqueIds), [])
  return {
    ...state,
    ...data,
    uniqueIds,
    films,
  }
}

export function appendFilms(state, data) {
  const newFilms = data.films || []
  const prevFilms = state.films || []
  const uniqueIds = state.uniqueIds
    || prevFilms.reduce((res, film) => {
      const id = film.imdbID

      if (!res[id]) {
        res[id] = id
      }
      return res
    }, {})
  return {
    ...state,
    ...data,
    uniqueIds,
    films: newFilms.reduce(reduceUniqueFilms.bind(null, uniqueIds), [
      ...prevFilms,
    ]),
  }
}
