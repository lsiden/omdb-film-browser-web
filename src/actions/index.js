import Actions from "./types"

export const updateFilmDetails = filmDetails => ({
  type: Actions.UPDATE_FILM_DETAILS,
  data: {
    filmDetails,
  },
})

export const updateFilms = (query = "", films = []) => ({
  type: Actions.UPDATE_FILMS,
  data: { films, query, pageNum: 1 },
})

export const appendFilms = (pageNum, films) => ({
  type: Actions.APPEND_FILMS,
  data: { films, pageNum },
})

export const updateIsFetching = isFetching => ({
  type: Actions.UPDATE_IS_FETCHING,
  data: { isFetching },
})
