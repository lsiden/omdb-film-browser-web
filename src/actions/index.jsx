import Actions from './types'

export const updateFilmDetails = (filmDetails) => ({
  type: Actions.UPDATE_FILM_DETAILS,
  data: {
    filmDetails,
  },
})

export const updateFilms = (query = '', films = [], totalResults = 0) => ({
  type: Actions.UPDATE_FILMS,
  data: {
    films, query, totalResults, pageNum: 1,
  },
})

export const appendFilms = (pageNum = 1, films = []) => ({
  type: Actions.APPEND_FILMS,
  data: { films, pageNum },
})

export const setQuery = (query = '') => ({
  type: Actions.SET_QUERY,
  data: { query },
})

export const setFetching = (isFetching = false) => ({
  type: Actions.UPDATE_IS_FETCHING,
  data: { isFetching }
})
