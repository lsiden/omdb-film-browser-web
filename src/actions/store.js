import { createStore, applyMiddleware, compose } from "redux"
import thunk from "redux-thunk"

import Actions from "actions/types"

const initialState = {
  view: Actions.VIEW_FILM_LIST,
}

export function reduce(state = initialState, action) {
  const { type, data } = action
  switch (type) {
    case Actions.VIEW_FILM_LIST:
    case Actions.VIEW_FILM_DETAIL:
    case Actions.UPDATE_FILMS:
    case Actions.UPDATE_FILM_DETAILS:
    case Actions.UPDATE_IS_FETCHING:
      return {
        ...state,
        ...data,
      }
    case Actions.APPEND_FILMS:
      return {
        ...state,
        ...data,
        films: [...state.films, ...data.films],
      }
    default:
      return state
  }
}

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
const store = createStore(reduce, composeEnhancers(applyMiddleware(thunk)))

export default store
