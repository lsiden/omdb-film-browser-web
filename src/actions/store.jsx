import { createStore, applyMiddleware, compose } from 'redux'
import thunk from 'redux-thunk'

import Actions from 'actions/types'
import { updateNewFilms, appendFilms } from 'actions/films'

const initialState = {
  view: Actions.VIEW_FILM_LIST,
  isFetching: false,
  films: [],
  totalResults: 0,
  query: '',
}

export function reduce(state = initialState, action) {
  const { type, data } = action
  switch (type) {
    case Actions.UPDATE_FILM_DETAILS:
    case Actions.UPDATE_IS_FETCHING:
    case Actions.SET_QUERY:
      return {
        ...state,
        ...data,
      }
    case Actions.UPDATE_FILMS:
      return updateNewFilms(state, data)
    case Actions.APPEND_FILMS:
      return appendFilms(state, data)
    default:
      return state
  }
}

const composeEnhancers = (
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({ trace: true, traceLimit: 25 })
  ) || compose
const store = createStore(reduce, composeEnhancers(applyMiddleware(thunk)))

export default store
