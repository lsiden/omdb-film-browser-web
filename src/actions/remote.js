// Remote Actions

import toastr from "toastr"

import { updateFilms, appendFilms, updateFilmDetails } from "./"
import store from "actions/store"

export const queryFetch = query => dispatch =>
  fetch(`https://www.omdbapi.com/?apikey=fbfcb8c7&type=movie&s=${query}`)
    .then(res => res.json())
    .then(res => dispatch(updateFilms(query, res.Search)))
    .catch(e => {
      console.error(e)
      toastr.error(e, "An error occured")
    })

export const pageFetch = () => dispatch => {
  const { query, pageNum = 1 } = store.getState()
  return fetch(
    `https://www.omdbapi.com/?apikey=fbfcb8c7&type=movie&s=${query}&page=${pageNum}`
  )
    .then(res => res.json())
    .then(res => dispatch(appendFilms(pageNum + 1, res.Search)))
    .catch(e => {
      console.error(e)
      toastr.error(e, "An error occured")
    })
}

export const fetchFilmDetails = id => dispatch =>
  fetch(`https://www.omdbapi.com/?apikey=fbfcb8c7&type=movie&i=${id}&plot=full`)
    .then(res => res.json())
    .then(res => dispatch(updateFilmDetails(res)))
    .catch(e => {
      console.error(e)
      toastr.error(e, "An error occured")
    })
