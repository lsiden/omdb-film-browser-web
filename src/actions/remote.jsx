// Remote Actions

import toastr from 'toastr'
import promiseFinally from 'promise.prototype.finally'

import store from 'actions/store'
import { FETCH_TIMEOUT } from 'omdb_constants'
import {
  updateFilms,
  appendFilms,
  updateFilmDetails,
  setFetching,
} from '.'

// promise.prototoype.finally is not yet available in node.js.
// This prevents tests from breaking.
// It is a no-op when .finally() is already defined.
promiseFinally.shim()

function toJson(res) {
  try {
    return res.json()
  } catch (e) {
    return Promise.reject(e)
  }
}

const onCatch = (e) => {
  console.error(e)
  toastr.error(e, 'An error occured', {
    preventDuplicates: true,
  })
}

// Returns Promise that resolves to fetch result if successful
const fetchWithTimeout = (dispatch, uri) => {
  let timeout
  const timeoutPromise = new Promise((_, reject) => {
    timeout = setTimeout(() => {
      reject(
        new Error(
          'Unable to fetch data. Try a different title or try again later.',
        ),
      )
    }, FETCH_TIMEOUT)
  })
  dispatch(setFetching(true))
  return Promise.race([timeoutPromise, fetch(uri)])
    .then(toJson)
    .catch(onCatch)
    .finally(() => {
      clearTimeout(timeout)
      dispatch(setFetching(false))
    })
}

class FetchCancelledError extends Error {
  constructor() {
    super()
    this.ignore = true
  }
}

let cancelPrevQueryFetch = () => {}

export const queryFetch = (query) => (dispatch) => {
  cancelPrevQueryFetch()

  return new Promise((resolve) => {
    cancelPrevQueryFetch = (_, reject) => {
      reject(new FetchCancelledError(`query "${query}" was cancelled`))
      dispatch(setFetching(false))
    }

    return fetchWithTimeout(
      dispatch,
      `https://www.omdbapi.com?apikey=fbfcb8c7&type=movie&s=${query}`,
    ).then((res = {}) => {
      if (res && res.Search && res.Search.length) {
        return resolve(
          dispatch(updateFilms(query, res.Search, res.totalResults)),
        )
      }
      return resolve(dispatch(updateFilms(query, [], 0)))
    })
  })
    .catch((e) => {
      if (!e.ignore) {
        console.error(e)
      }
    })
    .finally(() => {
      cancelPrevQueryFetch = () => {}
    })
}

// Returns a Promise
export const pageFetch = () => (dispatch) => {
  const { query, pageNum = 1 } = store.getState()
  const nextPageNum = pageNum + 1
  return fetchWithTimeout(
    dispatch,
    `https://www.omdbapi.com?apikey=fbfcb8c7&type=movie&s=${query}&page=${nextPageNum}`,
  ).then((res) => {
    if (res && res.Search && res.Search.length) {
      dispatch(appendFilms(nextPageNum, res.Search))
    }
  })
}

// Returns a Promise
export const fetchFilmDetails = (id) => (dispatch) => fetchWithTimeout(
  dispatch,
  `https://www.omdbapi.com?apikey=fbfcb8c7&type=movie&i=${id}&plot=full`,
).then((res) => dispatch(updateFilmDetails(res)))
