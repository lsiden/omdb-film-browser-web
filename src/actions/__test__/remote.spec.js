import { queryFetch, pageFetch, fetchFilmDetails } from "actions/remote"
import fetchMock from 'fetch-mock'

// Promise.prototype.catch = fn => new Promise((resolve, reject) => reject(fn))

fetchMock.config.overwriteRoutes = true

test("queryFetch()(dispatch) invokes dispatch({type: Actions.UPDATE_FILMS})", () => {
  const dispatch = jest.fn()
  const expectResponse = {
    Search: [
      { imdbID: 1, title: "Rocky Horror" },
      { imdbID: 2, title: "Halloween" },
    ],
  }
  fetchMock.mock('*', expectResponse)
  return queryFetch("a query")(dispatch).then(() => {
    expect(dispatch.mock.calls).toMatchSnapshot()
  })
})

test("pageFetch()(dispatch)", () => {
  const dispatch = jest.fn()
  const expectResponse = {
    Search: ["Rocky Horror 2", "Halloween 2"],
  }
  fetchMock.mock('*', expectResponse)
  return pageFetch()(dispatch).then(() => {
    expect(dispatch.mock.calls).toMatchSnapshot()
  })
})

test("fetchFilmDetails()(dispatch)", () => {
  const dispatch = jest.fn()
  const expectResponse = {
    filmDetails: {
      title: "a title",
      director: "Joe Director",
    },
  }
  fetchMock.mock('*', expectResponse)
  return fetchFilmDetails("id")(dispatch).then(() => {
    expect(dispatch.mock.calls).toMatchSnapshot()
  })
})
