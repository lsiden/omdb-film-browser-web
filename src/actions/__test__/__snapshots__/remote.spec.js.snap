// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchFilmDetails()(dispatch) 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "isFetching": true,
      },
      "type": "update-is-fetching",
    },
  ],
  Array [
    Object {
      "data": Object {
        "isFetching": false,
      },
      "type": "update-is-fetching",
    },
  ],
  Array [
    Object {
      "data": Object {
        "filmDetails": Object {
          "filmDetails": Object {
            "director": "Joe Director",
            "title": "a title",
          },
        },
      },
      "type": "update-film-details",
    },
  ],
]
`;

exports[`pageFetch()(dispatch) 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "isFetching": true,
      },
      "type": "update-is-fetching",
    },
  ],
  Array [
    Object {
      "data": Object {
        "isFetching": false,
      },
      "type": "update-is-fetching",
    },
  ],
  Array [
    Object {
      "data": Object {
        "films": Array [
          "Rocky Horror 2",
          "Halloween 2",
        ],
        "pageNum": 2,
      },
      "type": "append-films",
    },
  ],
]
`;

exports[`queryFetch()(dispatch) invokes dispatch({type: Actions.UPDATE_FILMS}) 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "isFetching": true,
      },
      "type": "update-is-fetching",
    },
  ],
  Array [
    Object {
      "data": Object {
        "isFetching": false,
      },
      "type": "update-is-fetching",
    },
  ],
  Array [
    Object {
      "data": Object {
        "films": Array [
          Object {
            "imdbID": 1,
            "title": "Rocky Horror",
          },
          Object {
            "imdbID": 2,
            "title": "Halloween",
          },
        ],
        "pageNum": 1,
        "query": "a query",
        "totalResults": 0,
      },
      "type": "update-films",
    },
  ],
]
`;
